here-naksha-storage-http/src/main/java/com/here/naksha/storage/http/RequestSender.java

```diff
@@ -0,0 +1,104 @@
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class RequestSender {
+
+  private static final Logger log = LoggerFactory.getLogger(RequestSender.class);
+  private final String hostUrl;
+
+  private final HttpClient httpClient;
+
+  // Do not modify after constructor because RequestSender is shared among multiple requests.
+  private final HttpRequest.Builder baseBuilder;
+  private final long socketTimeoutSec;
+
+  public RequestSender(
+      String hostUrl, Map<String, String> defaultHeaders, HttpClient httpClient, long socketTimeoutSec) {
+    this.hostUrl = hostUrl;
+    this.httpClient = httpClient;
+
+    this.baseBuilder = newBuilder().timeout(Duration.ofSeconds(socketTimeoutSec));
+    this.socketTimeoutSec = socketTimeoutSec;
+    defaultHeaders.forEach(baseBuilder::header);
+  }
+
+  /**
+   * Send a request configured based on enclosing {@link HttpStorage}.
+   *
+   * @param endpoint does not contain host:port part, starts with "/".
+   */
+  HttpResponse<String> sendRequest(String endpoint) {
+    return sendRequest(endpoint, null, null, null);
+  }
+
+  HttpResponse<String> sendRequest(
+      @NotNull String endpoint,
+      @Nullable Map<String, String> headers,
+      @Nullable String httpMethod,
+      @Nullable String body) {
+    URI uri = URI.create(hostUrl + endpoint);
+    HttpRequest.Builder specificBuilder = baseBuilder.copy().uri(uri);
+
+    if (headers != null) headers.forEach(specificBuilder::header);
+
+    HttpRequest.BodyPublisher bodyPublisher =
+        body == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(body);
+    if (httpMethod != null) specificBuilder.method(httpMethod, bodyPublisher);
+    HttpRequest specificRequest = specificBuilder.build();
+
+    return sendRequest(specificRequest);
+  }
+
+  private HttpResponse<String> sendRequest(HttpRequest specificRequest) {
+    try {
+      long startTime = System.currentTimeMillis();
+      CompletableFuture<HttpResponse<String>> futureResponse =
+          httpClient.sendAsync(specificRequest, HttpResponse.BodyHandlers.ofString());
+      HttpResponse<String> response = futureResponse.get(socketTimeoutSec, TimeUnit.SECONDS);
+      long executionTime = System.currentTimeMillis() - startTime;
+      log.info("Request to {} took {}ms", specificRequest.uri(), executionTime);
+      return response;
+    } catch (TimeoutException e) {
+      TimeoutException t = new TimeoutException(this.getClass().getName() + ": request timed out");
+      log.warn("Exception thrown.", t);
+      throw unchecked(t);
+    } catch (Exception e) {
+      log.warn("Exception thrown.", e);
```

Comment:
Will be good to put some context in log, saying something like "We got exception while executing Http request against remote server"

Reply:
Done